### PR TITLE
JDK-8301070: Replace NULL with nullptr in share/memory/

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi, this PR changes all occurrences of NULL to nullptr for the subdirectory share/memory/. Unfortunately the script that does the change isn't perfect, and so we
need to comb through these manually to make sure nothing has gone wrong. I also review these changes but things slip past my eyes sometimes.

Here are some typical things to look out for:

1. No changes but copyright header changed (probably because I reverted some changes but forgot the copyright).
2. Macros having their NULL changed to nullptr, these are added to the script when I find them. They should be NULL.
3. nullptr in comments and logs. We try to use lower case "null" in these cases as it reads better. An exception is made when code expressions are in a comment.

An example of this:

```c++
// This function returns null
void* ret_null();
// This function returns true if *x == nullptr
bool is_nullptr(void** x);
```

Note how `nullptr` participates in a code expression here, we really are talking about the specific value `nullptr`.

Thanks!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301070](https://bugs.openjdk.org/browse/JDK-8301070): Replace NULL with nullptr in share/memory/

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to [def9b5f3](https://git.openjdk.org/jdk/pull/12185/files/def9b5f323ae1f63d4e0f1942f3d3f1f93a01fff)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [def9b5f3](https://git.openjdk.org/jdk/pull/12185/files/def9b5f323ae1f63d4e0f1942f3d3f1f93a01fff)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12185/head:pull/12185` \
`$ git checkout pull/12185`

Update a local copy of the PR: \
`$ git checkout pull/12185` \
`$ git pull https://git.openjdk.org/jdk pull/12185/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12185`

View PR using the GUI difftool: \
`$ git pr show -t 12185`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12185.diff">https://git.openjdk.org/jdk/pull/12185.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/memory/heap.cpp [hunk 1]
```diff
@@ -644,40 +644,40 @@ void CodeHeap::add_to_freelist(HeapBlock* a) {
...
-  assert((prev < b) && (cur == NULL || b < cur), "free-list must be ordered");
+  assert((prev < b) && (cur == nullptr || b < cur), "free-list must be ordered");
   insert_after(prev, b);
   _last_insert_point = prev;
 }
 
 /**
  * Search freelist for an entry on the list with the best fit.
- * @return NULL, if no one was found
+ * @return nullptr, if no one was found
```
Comment (jdksjolen) [inline src/hotspot/share/memory/heap.cpp:666]: null

**Review by jdksjolen — COMMENTED** [review]
Found one manual fix

**Comment by jdksjolen** [comment]
Passes tier1.

File: src/hotspot/share/memory/guardedMemory.hpp [hunk 2]
```diff
@@ -226,7 +226,7 @@ class GuardedMemory : StackObj { // Wrapper on stack
   /**
    * Return the general purpose tag.
    *
-   * @return the general purpose tag, defaults to NULL.
+   * @return the general purpose tag, defaults to nullptr.
```
Comment (stefank) [inline src/hotspot/share/memory/guardedMemory.hpp:229]: nullptr => null?

File: src/hotspot/share/memory/guardedMemory.hpp [hunk 3]
```diff
@@ -192,8 +192,8 @@ class GuardedMemory : StackObj { // Wrapper on stack
    *
    * @return user data pointer (inner pointer to supplied "base_ptr").
    */
-  void* wrap_with_guards(void* base_ptr, size_t user_size, const void* tag = NULL) {
-    assert(base_ptr != NULL, "Attempt to wrap NULL with memory guard");
+  void* wrap_with_guards(void* base_ptr, size_t user_size, const void* tag = nullptr) {
+    assert(base_ptr != nullptr, "Attempt to wrap nullptr with memory guard");
```
Comment (stefank) [inline src/hotspot/share/memory/guardedMemory.hpp:196]: wrap nullptr => wrap null?

File: src/hotspot/share/memory/guardedMemory.hpp [hunk 4]
```diff
@@ -303,9 +303,9 @@ class GuardedMemory : StackObj { // Wrapper on stack
    * @param len the length of the copy
    * @param tag optional general purpose tag (see GuardedMemory::get_tag())
    *
-   * @return guarded wrapped memory pointer to the user area, or NULL if OOM.
+   * @return guarded wrapped memory pointer to the user area, or nullptr if OOM.
```
Comment (stefank) [inline src/hotspot/share/memory/guardedMemory.hpp:306]: nullptr => null?

**Review by stefank — APPROVED** [review]

File: src/hotspot/share/memory/metaspace/blockTree.hpp [hunk 5]
```diff
@@ -212,30 +212,30 @@ class BlockTree: public CHeapObj<mtMetaspace> {
...
 
   // Given a node n and an insertion point, insert n under insertion point.
   void insert(Node* insertion_point, Node* n) {
-    assert(n->_parent == NULL, "Sanity");
+    assert(n->_parent == nullptr, "Sanity");
     for (;;) {
       DEBUG_ONLY(check_node(insertion_point);)
       if (n->_word_size == insertion_point->_word_size) {
-        add_to_list(n, insertion_point); // parent stays NULL in this case.
+        add_to_list(n, insertion_point); // parent stays null in this case.
```
Comment (tstuefe) [inline src/hotspot/share/memory/metaspace/blockTree.hpp:228]: Small nit. Do we have a unified naming scheme for comments?
 
"null" reads like a java null. I prefer keeping NULL in comments.

**Review by tstuefe — APPROVED** [review]
Metaspace changes are good.

File: src/hotspot/share/memory/metaspace/blockTree.hpp (hunk 5 above)
Comment (jdksjolen) [reply src/hotspot/share/memory/metaspace/blockTree.hpp:228]: FWIW, from fixing these reviews I've never seen us referring to a Java null when we meant a C++ nullptr.

**Comment by jdksjolen** [comment]
/integrate

Passes tests and two approvals (with stefank's suggested fixes included), integrating.

**Pull request closed without merging**